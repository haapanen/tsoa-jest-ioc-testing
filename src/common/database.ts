import { Lifecycle, scoped } from "tsyringe";
import postgres from "postgres";

export class DatabaseConfig {
  constructor(
    public host: string,
    public port: string,
    public database: string,
    public username: string,
    public password: string
  ) {}
}

@scoped(Lifecycle.ContainerScoped)
export class Database {
  sql: postgres.Sql<{}>;

  constructor(private readonly databaseConfig: DatabaseConfig) {
    this.sql = postgres({
      host: databaseConfig.host,
      port: parseInt(databaseConfig.port),
      database: databaseConfig.database,
      username: databaseConfig.username,
      password: databaseConfig.password,
    });
  }
}
