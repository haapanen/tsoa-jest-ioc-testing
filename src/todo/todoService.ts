import { Lifecycle, scoped } from "tsyringe";
import { Database } from "../common/database";
import NotFoundError from "../common/errors/notFoundError";
import { Todo } from "./todo";

export type CreateTodoParameters = Pick<Todo, "description" | "done" | "title">;

@scoped(Lifecycle.ContainerScoped)
export class TodoService {
  constructor(private readonly database: Database) {}

  public async getTodos(): Promise<Todo[]> {
    const todos = await this.database.sql<Todo[]>`
        select 
            id,
            title,
            description,
            done
        from todos;
    `;

    return todos;
  }

  public async createTodo(parameters: CreateTodoParameters) {
    const [createdTodo] = await this.database.sql<Todo[]>`
        insert into todos ${this.database.sql(parameters)}
        returning id, title, description, done
    `;
  }

  public async updateTodo(todoId: number, todo: Todo) {
    const result = await this.database.sql`
        update todos set ${this.database.sql(
          todo,
          "title",
          "description",
          "done"
        )} where id = ${todoId} returning 1 as updated
    `;

    if (result.length === 0) {
      throw new NotFoundError(`No todo found with id ${todoId}`);
    }
  }
}
