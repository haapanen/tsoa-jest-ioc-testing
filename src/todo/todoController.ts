import { Body, Controller, Get, Path, Post, Put, Route, SuccessResponse } from "tsoa";
import { singleton } from "tsyringe";
import { Todo } from "./todo";
import { CreateTodoParameters, TodoService } from "./todoService";

@singleton()
@Route("api/v1/todos")
export class TodoController extends Controller {
  constructor(private readonly todoService: TodoService) {
    super();
  }

  @Get("")
  public async getTodos(): Promise<Todo[]> {
    return await this.todoService.getTodos();
  }

  @Post("")
  @SuccessResponse("201", "Created")
  public async createTodo(@Body() parameters: CreateTodoParameters): Promise<void> {
    this.setStatus(201);
    await this.todoService.createTodo(parameters);
  }

  @Put("{todoId}")
  @SuccessResponse("204", "Updated")
  public async updateTodo(@Path() todoId: number, @Body() todo: Todo): Promise<void> {
    this.setStatus(204);
    await this.todoService.updateTodo(todoId, todo);
  }
}
