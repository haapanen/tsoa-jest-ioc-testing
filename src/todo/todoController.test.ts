import "reflect-metadata";
import { Database } from "../common/database";
import { iocContainer } from "../ioc";
import { Todo } from "./todo";
import { TodoController } from "./todoController";

const removeTodos = async () => {
  const database = iocContainer.get(Database);

  await database.sql`delete from todos;`;
};

beforeEach(async () => {
  await removeTodos();
});

describe("Todos API", () => {
  it("Should create a todo", async () => {
    const todosController = iocContainer.get(TodoController);
    const database = iocContainer.get(Database);

    await todosController.createTodo({
      title: "title",
      description: "description",
      done: false,
    });

    const result = await database.sql<Todo[]>`select * from todos;`;

    expect(result[0]).toMatchSnapshot({
      id: expect.any(Number),
    });
  });
});
