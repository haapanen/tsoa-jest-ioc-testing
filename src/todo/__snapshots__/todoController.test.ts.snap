// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos API Should create a todo 1`] = `
Object {
  "description": "description",
  "done": false,
  "id": Any<Number>,
  "title": "title",
}
`;
